proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,199 @@
+* cannot contain a handle (such as resource types)
+
+Resource types are never allowed as RETs (i.e., `RWBuffer<int>` as an RET).
+Texture resources conform to the rules for typed buffers.
+If the target is SPIR-V, then only the set of rules for raw buffers apply. Typed buffers
+like `RWBuffer` may have RETs that exceed 128 bits, for example, if the target 
+IR is SPIR-V. The typed buffer rules above are only enforced when the IR target is DXIL.
+
+If someone writes `RWBuffer<MyCustomType>` and MyCustomType is not a 
+valid RET, there should be infrastructure to reject this RET and emit a message 
+explaining why it was rejected as an RET.
+
+## Motivation
+Currently, there is an allow list of valid RETs. It must be modified with respect 
+to this spec. Anything that is not a valid RET will be rejected. The allow list isn't
+broad enough, because it doesn't include the case where the RET is user-defined. 
+Ideally, a user should be able to determine how any user-defined structure is invalid 
+as an RET. Some system should be in place to more completely enforce the rules for 
+valid and invalid RETs, as well as provide useful information on why they are invalid.
+
+For example, when targeting DXIL IR, `RWBuffer<double4> b : register(u4);` will emit
+an error in DXC, but will not in clang-dxc, despite the fact that `double4` is an 
+invalid RET for typed buffers.
+
+## Proposed solution
+
+The proposed solution is to use C++20 concepts to validate the element type.
+A new built-in, `__builtin_hlsl_is_line_vector_layout_compatible`, will be added in 
+order to express constraints that can't currently be expressed in pure HLSL code.
+Standard clang diagnostics for unsatisfied constraints will be used to report any 
+invalid element types. Concepts required will differ depending on whether the 
+resource is a typed buffer or a raw buffer. Until concepts are formally supported 
+by HLSL, the concepts and constraints will be expressed directly in the AST via 
+the HLSL external sema source.
+
+## Detailed design
+
+In `clang\lib\Sema\HLSLExternalSemaSource.cpp`, `RWBuffer` is defined, along with 
+`RasterizerOrderedBuffer` and `StructuredBuffer`. It is at this point that the 
+type traits should be incorporated into these resource declarations. A concept 
+containing the relevant type traits will be applied to each resource declaration,
+and there wil be sufficient context to determine the target IR.
+If DXIL is given, then all of the typed buffer type traits will be applied on each
+typed buffer HLSL resource type. Otherwise, the raw buffer type traits will be 
+applied to each resource type. If a type trait is not true for the given 
+RET, a corresponding error message will be emitted.
+
+The list of type traits that will be available for use are described below:
+| type trait | Description|
+|-|-|
+| `__is_complete_type` | An RET should either be a complete type, or a user defined type that has been completely defined. |

```

Comment:
We may not actually need `__is_complete_type`. I know that the type needs to be complete, but `__is_intangible` will error if the type is incomplete.